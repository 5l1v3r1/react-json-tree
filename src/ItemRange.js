// @flow
import React from 'react';
import JSONArrow from './JSONArrow';

import type { NestedType, RenderChildNodes } from './types';
import type { StylingFunction } from 'react-base16-styling';

type Props = {
  styling: StylingFunction,
  from: number,
  to: number,
  renderChildNodes: RenderChildNodes,
  nodeType: NestedType
};

type State = {
  expanded: boolean
};

export default class ItemRange extends React.PureComponent<void, Props, State> {
  state = { expanded: false };

  render() {
    const { styling, from, to, renderChildNodes, nodeType } = this.props;

    return this.state.expanded
      ? <div {...styling(['itemRange', 'itemRangeColor'], this.state.expanded)}>
          {renderChildNodes(this.props, from, to)}
        </div>
      : <div
          {...styling(['itemRange', 'itemRangeColor'], this.state.expanded)}
          onClick={this.handleClick}
        >
          <JSONArrow
            nodeType={nodeType}
            styling={styling}
            expanded={false}
            onClick={this.handleClick}
            arrowStyle="double"
          />
          {`${from} ... ${to}`}
        </div>;
  }

  handleClick = () => {
    this.setState({ expanded: !this.state.expanded });
  };
}
