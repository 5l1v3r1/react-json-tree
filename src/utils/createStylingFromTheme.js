// @flow
import { createStyling } from 'react-base16-styling';
import jsontree from '../themes/jsontree';

const colorMap = theme => ({
  BACKGROUND_COLOR: theme.base00,
  TEXT_COLOR: theme.base07,
  STRING_COLOR: theme.base0B,
  DATE_COLOR: theme.base0B,
  NUMBER_COLOR: theme.base09,
  BOOLEAN_COLOR: theme.base09,
  NULL_COLOR: theme.base08,
  UNDEFINED_COLOR: theme.base08,
  FUNCTION_COLOR: theme.base08,
  SYMBOL_COLOR: theme.base08,
  LABEL_COLOR: theme.base0D,
  ARROW_COLOR: theme.base0D,
  ITEM_STRING_COLOR: theme.base0B,
  ITEM_STRING_EXPANDED_COLOR: theme.base03
});

const valueColorMap = colors => ({
  String: colors.STRING_COLOR,
  Date: colors.DATE_COLOR,
  Number: colors.NUMBER_COLOR,
  Boolean: colors.BOOLEAN_COLOR,
  Null: colors.NULL_COLOR,
  Undefined: colors.UNDEFINED_COLOR,
  Function: colors.FUNCTION_COLOR,
  Symbol: colors.SYMBOL_COLOR
});

const getDefaultThemeStyling = theme => {
  const colors = colorMap(theme);

  return {
    tree: {
      border: 0,
      padding: 0,
      margin: 0,
      listStyle: 'none',
      MozUserSelect: 'none',
      WebkitUserSelect: 'none',
      fontFamily: 'inconsolata, monospace'
    },

    treeColor: {
      backgroundColor: colors.BACKGROUND_COLOR
    },

    value: ({ style }, nodeType, keyPath) => ({
      style: {
        ...style,
        paddingTop: '0.25em',
        paddingLeft: keyPath.length > 1 ? '2.125em' : '1.25em',
        paddingRight: 0,
        marginLeft: '0.875em',
        WebkitUserSelect: 'text',
        MozUserSelect: 'text',
        wordWrap: 'break-word',
        textIndent: '-0.5em',
        wordBreak: 'break-all',
        whiteSpace: 'pre-wrap'
      }
    }),

    label: {
      display: 'inline-block'
    },

    labelColor: {
      color: colors.LABEL_COLOR
    },

    valueLabel: {
      margin: '0 0.5em 0 0'
    },

    valueTextColor: ({ style }, nodeType) => ({
      style: {
        ...style,
        color: valueColorMap(colors)[nodeType]
      }
    }),

    itemRange: ({ style }, expanded) => ({
      style: {
        ...style,
        paddingTop: expanded ? 0 : '0.25em',
        cursor: 'pointer'
      }
    }),

    itemRangeColor: {
      color: colors.LABEL_COLOR
    },

    arrow: ({ style }, nodeType, expanded) => ({
      style: {
        ...style,
        marginLeft: 0,
        transition: '150ms',
        WebkitTransition: '150ms',
        MozTransition: '150ms',
        WebkitTransform: expanded ? 'rotateZ(90deg)' : 'rotateZ(0deg)',
        MozTransform: expanded ? 'rotateZ(90deg)' : 'rotateZ(0deg)',
        transform: expanded ? 'rotateZ(90deg)' : 'rotateZ(0deg)',
        transformOrigin: '45% 50%',
        WebkitTransformOrigin: '45% 50%',
        MozTransformOrigin: '45% 50%',
        position: 'relative',
        lineHeight: '1.1em',
        fontSize: '0.75em'
      }
    }),

    arrowInner: {
      position: 'absolute',
      top: 0,
      left: '-0.4em'
    },

    arrowContainer: ({ style }, arrowStyle) => ({
      style: {
        ...style,
        display: 'inline-block',
        paddingRight: '0.5em',
        paddingLeft: arrowStyle === 'double' ? '1em' : 0,
        cursor: 'pointer'
      }
    }),

    arrowSignColor: {
      color: colors.ARROW_COLOR
    },

    nestedNode: ({ style }, keyPath, nodeType, expanded, expandable) => ({
      style: {
        ...style,
        position: 'relative',
        paddingTop: keyPath.length > 1 ? '0.25em' : 0,
        marginLeft: keyPath.length > 1 ? '0.875em' : 0,
        paddingLeft: !expandable ? '1.125em' : 0
      }
    }),

    rootNode: {
      padding: 0,
      margin: 0
    },

    nestedNodeLabel: ({ style }, keyPath, nodeType, expanded, expandable) => ({
      style: {
        ...style,
        margin: 0,
        padding: 0,
        WebkitUserSelect: expandable ? 'inherit' : 'text',
        MozUserSelect: expandable ? 'inherit' : 'text',
        cursor: expandable ? 'pointer' : 'default'
      }
    }),

    nestedNodeItemPreview: {
      paddingLeft: '0.5em',
      cursor: 'default'
    },

    nestedNodeItemPreviewColor: ({ style }, keyPath, nodeType, expanded) => ({
      style: {
        ...style,
        color: expanded
          ? colors.ITEM_STRING_EXPANDED_COLOR
          : colors.ITEM_STRING_COLOR
      }
    }),

    nestedNodeChildren: ({ style }, nodeType, expanded) => ({
      style: {
        ...style,
        padding: 0,
        margin: 0,
        listStyle: 'none',
        display: expanded ? 'block' : 'none'
      }
    }),

    rootNodeChildren: {
      padding: 0,
      margin: 0,
      listStyle: 'none'
    }
  };
};

export default createStyling(getDefaultThemeStyling, {
  defaultBase16: jsontree,
  base16Themes: { jsontree }
});
