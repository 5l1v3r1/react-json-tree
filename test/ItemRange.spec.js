// @flow
import React from 'react';
import test from 'ava';
import renderer from 'react-test-renderer';

import ItemRange from '../src/ItemRange';

import createStylingFromTheme from '../src/utils/createStylingFromTheme';

const styling = createStylingFromTheme(undefined);

test('should render ItemRange', t => {
  const tree = renderer.create(
    <ItemRange
      styling={styling}
      from={0}
      to={10}
      renderChildNodes={(from, to) => null}
      nodeType="Array"
    />
  );
  t.snapshot(tree.toJSON());
});
