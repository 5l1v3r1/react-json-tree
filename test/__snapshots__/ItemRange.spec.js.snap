exports[`should render ItemRange 1`] = `"{\"__ava_react_jsx\":{\"type\":\"div\",\"props\":{\"style\":{\"paddingTop\":\"0.25em\",\"cursor\":\"pointer\",\"color\":\"#2287cd\"}},\"children\":[{\"type\":\"div\",\"props\":{\"style\":{\"display\":\"inline-block\",\"paddingRight\":\"0.5em\",\"paddingLeft\":\"1em\",\"cursor\":\"pointer\"}},\"children\":[{\"type\":\"div\",\"props\":{\"style\":{\"marginLeft\":0,\"transition\":\"150ms\",\"WebkitTransition\":\"150ms\",\"MozTransition\":\"150ms\",\"WebkitTransform\":\"rotateZ(0deg)\",\"MozTransform\":\"rotateZ(0deg)\",\"transform\":\"rotateZ(0deg)\",\"transformOrigin\":\"45% 50%\",\"WebkitTransformOrigin\":\"45% 50%\",\"MozTransformOrigin\":\"45% 50%\",\"position\":\"relative\",\"lineHeight\":\"1.1em\",\"fontSize\":\"0.75em\",\"color\":\"#2287cd\"}},\"children\":[\"▶\",{\"type\":\"div\",\"props\":{\"style\":{\"position\":\"absolute\",\"top\":0,\"left\":\"-0.4em\",\"color\":\"#2287cd\"}},\"children\":[\"▶\"]}]}]},\"0 ... 10\"]}}"`;
